From: Branden Robinson <branden@debian.org>
Subject: Add RemovePid() function, and register it with atexit()

Debian bug#213076

Forward-ported by David Nusinow, Eugene Konev and Julien Cristau.

Index: xdm/dm.c
===================================================================
--- xdm.orig/dm.c
+++ xdm/dm.c
@@ -107,6 +107,7 @@
 #endif
 
 static int StorePid (void);
+static void RemovePid (void);
 
 static pid_t parent_pid = -1; 	/* PID of parent xdm process */
 
@@ -162,6 +163,9 @@
 
     LogInfo ("Starting\n");
 
+    if (atexit (RemovePid))
+	LogError ("could not register RemovePid() with atexit()\n");
+
     if (nofork_session == 0) {
 	/* Clean up any old Authorization files */
 	/* AUD: all good? */
@@ -930,6 +934,19 @@
     return 0;
 }
 
+/*
+ * Remove process ID file.  This function is registered with atexit().
+ */
+static void
+RemovePid (void)
+{
+    Debug ("unlinking process ID file %s\n", pidFile);
+    if (unlink (pidFile))
+	if (errno != ENOENT)
+	    LogError ("cannot remove process ID file %s: %s\n", pidFile,
+		      _SysErrorMsg (errno));
+}
+
 #if 0
 void
 UnlockPidFile (void)
